os/board/rtl8730e/src/rtl8730e_alc1019.c

```diff
@@ -0,0 +1,211 @@
+ * Pre-processor Definitions
+ ****************************************************************************/
+/* i2c config */
+#define ALC1019_I2C_PORT		1
+#define ALC1019_I2C_FREQ		100000
+#define ALC1019_I2C_ADDRLEN		7
+
+#define ALC1019_I2C_ADDR		0x28 
+
+/* i2s config */
+#define ALC1019_I2S_PORT		2
+#define ALC1019_I2S_IS_MASTER		1
+#define ALC1019_I2S_IS_OUTPUT		1
+
+/*other pin config */
+#define ALC1019_GPIO_RESET_PIN		PA_22
+
+#define ALC1019_AVAILABLE_MINOR_MIN	0
+#define ALC1019_AVAILABLE_MINOR_MAX	25
+
+#ifdef CONFIG_AUDIO_ALC1019
+
+extern struct i2s_dev_s *amebad_i2s_initialize(uint16_t port);
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+struct rtl8730e_alc1019_audioinfo_s {
+	struct alc1019_lower_s lower;
+	alc1019_handler_t handler;
+	FAR void *arg;
+	gpio_t reset;
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+static void alc1019_control_reset_pin(bool active);
+static void alc1019_control_powerdown(bool powerdown);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+static struct rtl8730e_alc1019_audioinfo_s g_alc1019info = {
+	.lower = {
+		.i2c_config = {
+			.frequency = ALC1019_I2C_FREQ,
+			.address = ALC1019_I2C_ADDR,
+			.addrlen = ALC1019_I2C_ADDRLEN,
+		},
+	},
+	.reset.pin = ALC1019_GPIO_RESET_PIN,
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static void alc1019_control_reset_pin(bool active)
+{
+	gpio_write(&(g_alc1019info.reset), !active);
+}
+
+static void alc1019_control_powerdown(bool enter_powerdown)
+{
+	/* currently does nothing */
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: cy4390x_alc1019_initialize
+ *
+ * Description:
+ *   This function is called by platform-specific, setup logic to configure
+ *   and register the ALC1019 device.  This function will register the driver
+ *   as /dev/alc1019[x] where x is determined by the minor device number.
+ *
+ * Input Parameters:
+ *   minor - The input device minor number
+ *
+ * Returned Value:
+ *   Zero is returned on success.  Otherwise, a negated errno value is
+ *   returned to indicate the nature of the failure.
+ *
+ ****************************************************************************/
+int rtl8730e_alc1019_initialize(int minor)
```

Comment:
What are rtl8730e specific codes? Except the port number, the function looks common.

Reply:
Because some frequency of i2c and some more setting need to be changed based on attached amp or dsp ic, this style is better for readability. Otherwise i2c & i2s & spi of dsp should have some specific dev structure based on attached ic with more and more defconfig.